components/service-operator/internal/aws/cloudformation/controller.go

```diff
@@ -0,0 +1,332 @@
+// outputs defined in their template AND implement object.SecretNamer will
+// result in a secret being created.
+func (r *Controller) updateCredentialsSecret(ctx context.Context, o Stack, outputs Outputs) error {
+	if len(outputs) == 0 {
+		// no outputs to write to secret
+		return nil
+	}
+	s, ok := o.(object.SecretNamer)
+	if !ok {
+		// this objects is not a SecretNamer and does not
+		// require writing credentials to a secret
+		return nil
+	}
+	secret := core.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      s.GetSecretName(),
+			Namespace: o.GetNamespace(),
+		},
+		Data: map[string][]byte{},
+	}
+	secretKey, err := client.ObjectKeyFromObject(&secret)
+	if err != nil {
+		return err
+	}
+	err = r.KubernetesClient.Get(ctx, secretKey, &secret)
+	if err != nil && !apierrs.IsNotFound(err) {
+		return err
+	}
+	op, err := controllerutil.CreateOrUpdate(ctx, r.KubernetesClient, &secret, func() error {
+		secret.Type = core.SecretTypeOpaque
+		secret.Annotations = map[string]string{
+			"operator": "gsp-service-operator",
+			"group":    o.GroupVersionKind().Group,
+			"version":  o.GroupVersionKind().Version,
+		}
+		for key, value := range outputs {
+			secret.Data[key] = []byte(value)
+		}
+		return nil
+	})
+	r.Log.Info("update-secret",
+		"secret", secretKey,
+		"op", op,
+		"err", err,
+	)
+	if err != nil {
+		return err
+	}
+	// mark the secret as owned by the o resource so it gets gc'd
+	if err := controllerutil.SetControllerReference(o, &secret, r.Scheme); err != nil {
+		return err
+	}
+	return nil
+}
+
+// getRoleParams fetches additional IAM role params from any Stack that also
+// implements StackPolicyAttacher.  Implementing StackPolicyAttacher tells the
+// controller that this Stack wants to attach a policy to an IAM role, and
+// provides the methods to locate the object that represents the role and
+// extract the required role name.
+// FIXME: this method needs to know too much about "Principal". The presence of
+//        the PrincipalListKind var , PrincipalLister and Principal interface
+//        types hints that this abstraction is failing. which leads to some
+//        over complicated implementation of this controller. We should
+//        consider either: (1) dropping the "Prinipal" concept and letting this
+//        controller target roles based on annotation, (simple but may not work
+//        well with multiple CLOUD_PROVIDER values in the future OR (2)
+//        targeting "Pods" with a label selector instead of "Principal", and
+//        letting a controller manage both the creation of IAMRoles and the
+//        assignment of kiam annotations which would mean this controller would
+//        only need to know instead an IAMPolicy kubernetes object and have a
+//        policy OR (3) something else
+func (r *Controller) getRoleParams(ctx context.Context, o Stack) ([]*Parameter, error) {
+	serviceThatRequiresIAM, ok := o.(StackPolicyAttacher)
+	if !ok {
+		return nil, nil
```

Comment:
How about I move the check of satisfying the interface to before the func call? ... I feel this would illustrate the intension (that only things that assign policies need to request a role name) more clearly?